src/python-common/ceph/deployment/drive_group.py

```diff
@@ -106,6 +106,24 @@ def __init__(self, msg):
         super(DriveGroupValidationError, self).__init__('Failed to validate Drive Group: ' + msg)
 
 
+class DriveGroupSpecs(object):
+    """ Container class to parse drivegroups """
+
+    def __init__(self, drive_group_json):
```

Comment:
Hm. If the dashboard wants to create a `DriveGroupSpecs`, it needs to assemble the Drive Groups and then convert it to JSON to be possible to pass it into __init__? Why not make `DriveGroupSpecs.__init__` accept the `drive_groups` directly?

Reply:
Originally this is due the orchestrator cli passing only json/yaml, hence I just handed it over, but you're right. The dashboard call call directly in this function without the detour via the orch. I'll add support for json _and_ `DriveGroupSpec`

Reply:
but are we sure that the dashboard will actually assemble the drivegroups themselves? How fancy will the interface be? It may just assemble json/yaml and pass it to the orchestrator. maybe @bk201 knows?

Reply:
The frondend formats single dg spec to JSON and send it to backend. Backend then deserialize it. https://github.com/ceph/ceph/blob/8e00a36d5339fe8ee5131e8c128c593259a66313/src/pybind/mgr/dashboard/controllers/orchestrator.py#L108

Not sure how to use multiple dg in single request yet.

Reply:
Turns out you're right:

https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/controllers/orchestrator.py#L104-L110

on the other hand, all data structures in the orchestrator.py have the same interface:

1. `__init__` behaves compatible to [Data Classes](https://docs.python.org/3/library/dataclasses.html) in order to migrate to them as soon as we're on Py3
2. `from_json` to additionally accept serialized data 

I wouldn't want to introduce an exception here. So either we stay with this `from_json` interface, or we refactor everything to make `__init__` accept serialized data.

Reply:
1) as soon as we're on py3.8 or later, where dataclasses will be introduced.. I fear a majority of the supported platforms will be on 3.6/7 for quite some time.